tests/unit/test_upload.py

```diff
@@ -262,6 +266,71 @@ def test_stores_bodies_in_memory_upload_part(self):
             self.upload_input_manager.stores_body_in_memory('upload_part'))
 
 
+class TestUploadNonSeekableInputManager(TestUploadFilenameInputManager):
+    def setUp(self):
+        super(TestUploadNonSeekableInputManager, self).setUp()
+        self.upload_input_manager = UploadNonSeekableInputManager(
+            self.osutil, self.transfer_coordinator)
+        self.fileobj = NonSeekableReader(self.content)
+        self.call_args = CallArgs(
+            fileobj=self.fileobj, subscribers=self.subscribers)
+        self.future = self.get_transfer_future(self.call_args)
+
+    def assert_multipart_parts(self):
+        self.assertTrue(
+            self.upload_input_manager.requires_multipart_upload(
+                self.future, self.config))
+
+        parts = list(self.upload_input_manager.yield_upload_part_bodies(
+            self.future, self.config))
+
+        size = self.config.multipart_chunksize
+        num_parts = len(self.content) / size
+        num_parts = math.ceil(num_parts)
+        self.assertEqual(len(parts), num_parts)
+
+        for i, part in enumerate(parts[:-1]):
```

Comment:
I can follow this but it took a little bit of time. Comments would have helped especially in the for loop, the initial part calculations, and the final part assertion.